keps/sig-api-machinery/0006-apply.md

```diff
@@ -213,6 +223,226 @@ Kubernetes clusters. The semantical differences between server-side
+serialization problem needs to be solved.
+
+##### FieldsV1
+
+The current fieldset format, FieldsV1, is defined as a trie,
+serialized in JSON. Each key in the trie is either a ‘.’, or a
+serialized path element. If the key is a ‘.’, this represents the
+inclusion of the path leading to this key in the fieldset, and will
+always map to an empty set. If the key is a serialized path element,
+it represents a sub-field or item, and is split into two parts by the
+separating character ‘:’. The character before the separator is a type
+prefix and defines the format of the content after the separator,
+which will follow one of these four formats:
+
+  * 'f:<name>', where <name> is the name of a field in a struct, or
+  * 'key in a map v:<value>', where <value> is the exact json
+  * 'formatted value of a list item i:<index>', where <index> is
+  * 'position of an item in a list k:<keys>', where <keys> is a map of
+  * ' a list item's key fields to their unique values
+
+If a path element maps to an empty set, this is a shortcut for mapping
+to {‘.’}
+
+This format was chosen partly because it can be stored as a
+RawExtension in the API, which allows it to display in json or yaml
+depending on what a client asks for, which makes it more human
+readable. But because of this there are also some major inefficiencies
+to this format. For example, being serialized json means that we are
+wasting a lot of space on control characters, for example, to
+represent the list item with the value ‘udp’ for it’s field
+‘protocol’, we need to serialize `{“k:{\”protocol\”:\”udp\”}”:{}}`,
+which is a total of 29 characters, 9 are because it’s serialized JSON,
+11 are just for the format of the path element, 8 are to store the
+field name, and only 3 are for identifying the list item uniquely.
+
+##### Areas of Improvement
+
+  1. The first round of improvements, proposed in this
+  [PR](https://github.com/kubernetes-sigs/structured-merge-diff/pull/102),
+  restructures the trie from nested maps to nested lists, because it
+  permits directly serializing keys rather than encoding them as
+  strings, which wastes much less space double-encoding strings. It
+  also groups the path element type prefixes and whether or not self
+  is included into a single int, by taking a path element type int (f
+  -> 0, v -> 1, i -> 2, k -> 3) and adding either 0, 4, or 8 to
+  specify inclusion of self, children, or both respectively. These
+  improvements reduce the serialized size by about 22%, but there are
+  still some obvious areas we can improve.
+
+  1. One is field names, which are now a large part of the size of the
+  data, since we still directly serialize them as a string. To reduce
+  the size further we will implement a statically defined string
+  table, which maps int to strings, and instead of directly
+  serializing common kubernetes field names, we will just write their
+  index in this string table in base64, with a prefix of ! to specify
```

Comment:
We should have an escaping rule for if a unencoded string value is prefixed with `!`. I realize it is not needed for field names, but this encoder might be used for more than just field names, and it's a cheap escape check since it's just the first char in each string.

Reply:
I don't understand what the escaping rule would protect against? "!" is not valid JSON, deliberately chosen to denote a string table entry?

Reply:
I wasn't clear on if we're going to maintain a modified JSON grammer or just doing direct string substitution on a []byte. If we're maintaining the modified grammar, I agree escaping is not needed.

Reply:
Ah, I see what you're saying. Yeah, I was envisioning this step being done via (e.g.) an alternative implementation of jsoniter's reading/writing constructs. Which is implicitly making a modified grammar.